### PR TITLE
Don't block on peer recovery on the target side

### PR DESCRIPTION
**Pull request by s1monw**
Today we block using the generic thread-pool on the target side
until the source side has fully executed the recovery. We still
block on the source side executing the recovery in a blocking fashion
but there is no reason to block on the target side. This will
release generic threads early if there are many concurrent recoveries
happen.

Relates to #36195

### CODE REVIEW

**Review by ywelsch — CHANGES_REQUESTED** [review]
The response handler will need to run under GENERIC as both the happy and the retry/fail path do expensive blocking calls. As a follow-up, we can investigate whether we can get rid of `RecoveryTarget.cancellableThreads`.

File: server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java [hunk 1]
```diff
@@ -189,51 +192,11 @@ private void doRecovery(final long recoveryId) {
                 return;
             }
         }
-
-        try {
-            logger.trace("{} starting recovery from {}", request.shardId(), request.sourceNode());
-            final AtomicReference<RecoveryResponse> responseHolder = new AtomicReference<>();
-            cancellableThreads.execute(() -> responseHolder.set(
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java]: The [declaration and assignment of `cancellableThreads` above](https://github.com/elastic/elasticsearch/pull/37076/files#diff-cdd186e2c33a5fa4d4200f9119684443L177) is no longer used.

File: server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java [hunk 2]
```diff
@@ -189,51 +192,11 @@ private void doRecovery(final long recoveryId) {
...
-                        .append(", took [").append(timeValueMillis(recoveryResponse.phase2Time)).append("]")
-                        .append("\n");
-                logger.trace("{}", sb);
-            } else {
-                logger.debug("{} recovery done from [{}], took [{}]", request.shardId(), request.sourceNode(), recoveryTime);
-            }
-        } catch (CancellableThreads.ExecutionCancelledException e) {
-            logger.trace("recovery cancelled", e);
-        } catch (Exception e) {
+        Consumer<Throwable> handleException = e -> {
```
[the remark this answers is not in the document]
Comment (s1monw) [reply server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java]: unfortunately we need Throwable here - some consumer below needs it

File: server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java [hunk 3]
```diff
@@ -285,6 +250,61 @@ public RecoveryResponse read(StreamInput in) throws IOException {
...
+
+        try {
+            logger.trace("{} starting recovery from {}", request.shardId(), request.sourceNode());
+            transportService.submitRequest(request.sourceNode(), PeerRecoverySourceService.Actions.START_RECOVERY, request,
+                new TransportResponseHandler<RecoveryResponse>() {
+                    @Override
+                    public void handleResponse(RecoveryResponse recoveryResponse) {
+                        final TimeValue recoveryTime = new TimeValue(timer.time());
+                        // do this through ongoing recoveries to remove it from the collection
+                        onGoingRecoveries.markRecoveryAsDone(recoveryId);
```
[the remark this answers is not in the document]
Comment (s1monw) [reply server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java]: ++

File: server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java]: There are two consumers, one passes in an Exception, the other one a TransportException (which is also an Exception).

[the remark this answers is not in the document]
Comment (s1monw) [reply server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java]: nevermind I fixed it

File: server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java (hunk 1 above)
Comment (s1monw) [reply server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java]: ++ thanks

**Comment by s1monw** [comment]
@DaveCTurner @ywelsch I pushed changes

**Comment by s1monw** [comment]
folks I added back the cancelableThreads for now, I am not 100% convinced it's not needed and I would like to investigate it in a followup. I will also try to remove it entirely from the `RecoveryTarget` as @ywelsch  suggested. Yet, give how delicate the place of this call is I would like to only do it in 7.0 and only if we don't see issues backport it maybe later.

**Review by ywelsch — APPROVED** [review]
LGTM.

> Yet, give how delicate the place of this call is I would like to only do it in 7.0 and only if we don't see issues backport it maybe later

++

**Pull request merged**